domain/network/state/linklayer_merge.go

```diff
@@ -768,3 +818,155 @@ SET subnet_uuid = (
+	}
+
+	result := transform.SliceToMap(output, func(in providerSubnetCIDR) (string, providerSubnetCIDR) {
+		return in.ProviderID, in
+	})
+
+	return result, nil
+}
+
+// ensureAddressSubnetSuffix return a address including the subnet's suffix
+// and the subnetUUID. If the address is not an IPv4 nor IPv6 address, no
+// suffix is added.
+func ensureAddressSubnetSuffix(value, subnetProviderID string, data map[string]providerSubnetCIDR) (string, string, error) {
+	// Getting the subnet is helpful, but not essential.
+	subnet, _ := data[subnetProviderID]
+
+	parts := strings.Split(value, "/")
+	if len(parts) == 2 {
+		// Best case, the address has a CIDR suffix.
+		return subnet.SubnetUUID, value, nil
+	}
+
+	// Find the CIDR suffix, try the subnetCIDR first, if
+	// not define based on address type.
+	var suffix string
+	subnetCIDRParts := strings.Split(subnet.CIDR, "/")
+	switch len(subnetCIDRParts) {
+	case 2:
+		suffix = subnetCIDRParts[1]
+	case 1:
+		addType := corenetwork.DeriveAddressType(value)
+		if addType == corenetwork.IPv4Address {
+			suffix = "/32"
+		} else if addType == corenetwork.IPv6Address {
+			suffix = "/128"
+		} else {
+			return subnet.SubnetUUID, value, nil
+		}
+	case 0:
+		return "", "", errors.Errorf("invalid subnet CIDR value: %q", subnet.CIDR)
+	}
+
+	addressValue := value + "/" + suffix
+	return subnet.SubnetUUID, addressValue, nil
+}
+
+func (st *State) addAddressFromProvider(ctx context.Context, tx *sqlair.TX, add map[string][]mergeAddress) error {
+	if len(add) == 0 {
+		return nil
+	}
+
+	subnets, err := st.subnetCIDRUUIDByProviderID(ctx, tx, add)
+	if err != nil {
+		return errors.Capture(err)
+	}
+	lookups, err := st.getNetConfigLookups(ctx, tx)
+	if err != nil {
+		return errors.Capture(err)
+	}
+	providerIDAddrs := make(map[string]string)
+	ipAddresses := make([]ipAddress, 0)
+	for devUUID, addresses := range add {
+		for _, addr := range addresses {
+			ipAddrUUID, err := uuid.NewUUID()
+			if err != nil {
+				return errors.Capture(err)
+			}
+			subnetUUID, value, err := ensureAddressSubnetSuffix(addr.Value, addr.ProviderSubnetID, subnets)
+			if err != nil {
+				return errors.Capture(err)
+			}
+			var (
+				typeID, configID, scopeID, originID int
+			)
+			typeID = lookups.addrType[addr.AddressType]
+			configID = lookups.addrConfigType[addr.ConfigType]
+			scopeID = lookups.scope[addr.Scope]
+			originID = lookups.origin[addr.Origin]
+			ipAddresses = append(ipAddresses, ipAddress{
+				UUID:         ipAddrUUID.String(),
+				DeviceUUID:   devUUID,
+				AddressValue: value,
+				SubnetUUID:   nilZeroPtr(subnetUUID),
+				Type:         typeID,
+				ConfigType:   configID,
+				Origin:       originID,
+				Scope:        scopeID,
+				IsSecondary:  addr.IsSecondary,
+				IsShadow:     addr.IsShadow,
+			})
+			providerIDAddrs[addr.ProviderID] = ipAddrUUID.String()
+		}
+	}
+	// ensure address value subnet mask.
+	insertAddressStmt, err := sqlair.Prepare(`
+INSERT INTO ip_address
+SELECT
```

Comment:
This doesn't need to be a sub-select on account of the node. This whole merge operation is for a known node. Just pass it through.